google/field_helpers.go

```diff
@@ -5,41 +5,85 @@ import (
+}
 
-var networkLinkRegex = regexp.MustCompile("projects/(.+)/global/networks/(.+)")
+// ------------------------------------------------------------
+// Base helpers used to create helpers for specific fields.
+// ------------------------------------------------------------
 
-type NetworkFieldValue struct {
+type GlobalFieldValue struct {
 	Project string
 	Name    string
+
+	resourceType string
+}
+
+func (f GlobalFieldValue) RelativeLink() string {
+	if len(f.Name) == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf(globalLinkTemplate, f.Project, f.resourceType, f.Name)
 }
 
-// Parses a `network` supporting 5 different formats:
-// - https://www.googleapis.com/compute/{version}/projects/myproject/global/networks/my-network
-// - projects/myproject/global/networks/my-network
-// - global/networks/my-network (default project is used)
-// - my-network (default project is used)
-// - "" (empty string). RelativeLink() returns empty. For most API, the behavior is to use the default network.
-func ParseNetworkFieldValue(network string, config *Config) *NetworkFieldValue {
-	if networkLinkRegex.MatchString(network) {
-		parts := networkLinkRegex.FindStringSubmatch(network)
-
-		return &NetworkFieldValue{
+// Parses a global field supporting 4 different formats:
+// - https://www.googleapis.com/compute/ANY_VERSION/projects/{my-project}/global/{resource_type}/{resource_name}
+// - projects{my-project}/global/{resource_type}/{resource_name}
+// - global/{resource_type}/{resource_name} (default project is used)
+// - resource_name (default project is used)
+// - "" (empty string). RelativeLink() returns empty if isEmptyValid is true.
+func parseGlobalFieldValue(resourceType, fieldValue, projectSchemaField string, d TerraformResourceData, config *Config, isEmptyValid bool) (*GlobalFieldValue, error) {
+	if len(fieldValue) == 0 {
+		if isEmptyValid {
+			return &GlobalFieldValue{resourceType: resourceType}, nil
+		}
+		return nil, fmt.Errorf("The global field for resource %s cannot be empty", resourceType)
+	}
+
+	r := regexp.MustCompile(fmt.Sprintf(globalLinkBasePattern, resourceType))
+
+	if r.MatchString(fieldValue) {
+		parts := r.FindStringSubmatch(fieldValue)
+
+		return &GlobalFieldValue{
 			Project: parts[1],
 			Name:    parts[2],
-		}
+
+			resourceType: resourceType,
+		}, nil
 	}
 
-	return &NetworkFieldValue{
-		Project: config.Project,
-		Name:    GetResourceNameFromSelfLink(network),
+	project, err := getProjectFromSchema(projectSchemaField, d, config)
+	if err != nil {
+		return nil, err
 	}
+
+	return &GlobalFieldValue{
+		Project: project,
+		Name:    GetResourceNameFromSelfLink(fieldValue),
+
+		resourceType: resourceType,
+	}, nil
 }
 
-func (f NetworkFieldValue) RelativeLink() string {
-	if len(f.Name) == 0 {
-		return ""
+func getProjectFromSchema(projectSchemaField string, d TerraformResourceData, config *Config) (string, error) {
+	res, ok := d.GetOk(projectSchemaField)
```

Comment:
Are there any resources where this isn't just "project"?

Reply:
In `google_compute_instance` and `google_compute_instance_template`, there is a `subnetwork_project` field. This is the project specifically for the subnetwork and not the instance itself (with cross project networking, it is possible).